File: src/main/java/br/com/tech/challenge/service/SessionService.java
package br.com.tech.challenge.service;


import br.com.tech.challenge.domain.SessionDefinition;
import br.com.tech.challenge.exceptions.SessionAlreadyExistsException;
import br.com.tech.challenge.exceptions.SessionNotFoundException;
import br.com.tech.challenge.model.Session;
import br.com.tech.challenge.repository.SessionRepository;
import jakarta.transaction.Transactional;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.AmqpException;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.core.MessageBuilder;
import org.springframework.amqp.core.MessagePostProcessor;
import org.springframework.amqp.core.MessageProperties;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.UUID;

import static br.com.tech.challenge.config.SessionRabbitMQConfiguration.DELAY_HEADER;
import static br.com.tech.challenge.config.SessionRabbitMQConfiguration.ROUTING_KEY;
import static br.com.tech.challenge.config.SessionRabbitMQConfiguration.TOPIC_EXCHANGE_NAME;

@Service
@AllArgsConstructor
@Slf4j
public class SessionService {
    private final SessionRepository sessionRepository;
    private final RabbitTemplate rabbitTemplate;
    private final ResultService resultService;
    private final TopicService topicService;

    public Session create(SessionDefinition sessionDefinition) {
        if (sessionRepository.existSessionForTopic(sessionDefinition.topic().getId())) {
            throw new SessionAlreadyExistsException(STR."There is already a session for topic {ID} = \{sessionDefinition.topic().getId()}");
        }
        LocalDateTime now = LocalDateTime.now();
        Session session = Session.builder()
                .topic(sessionDefinition.topic())
                .startedAt(now)
                .finishAt(now.plusMinutes(sessionDefinition.timeToVote()))
                .build();

        session = save(session);
        publishSessionToClose(session);
        return session;
    }

    private void publishSessionToClose(Session session) {
        Message message = MessageBuilder.withBody(session.getId().toString().getBytes())
                .setContentType(MessageProperties.CONTENT_TYPE_JSON)
                .setHeader(DELAY_HEADER, session.timeToClose() * 1000)
                .build();
        rabbitTemplate.convertAndSend(TOPIC_EXCHANGE_NAME, ROUTING_KEY, message);
    }

    public Session save(Session session) {
        return sessionRepository.save(session);
    }

    @Transactional
    public void closeSession(String sessionId) {
        log.info(STR."Closing session for id {ID} " + sessionId);
        var session = sessionRepository.findById(UUID.fromString(sessionId)).orElseThrow(() ->
                new SessionNotFoundException(STR."There is no session open for id {ID} " + sessionId));
        resultService.buildResultDetailForSessionId(session.getId());
        var topic = session.getTopic();
        topic.setResult(session.result());
        topicService.save(topic);
    }

    @Cacheable("sessions")
    public Session findByTopicId(String id) {
        return sessionRepository.findByTopicId(UUID.fromString(id)).orElseThrow(() ->
                new SessionNotFoundException(STR."There is no session open for topic {ID} " + id));
    }

}

File: src/main/java/br/com/tech/challenge/service/SessionService.kt
package br.com.ntconsult.challenge.service

import br.com.ntconsult.challenge.domain.SessionDefinition
import br.com.ntconsult.challenge.exception.SessionAlreadyExistsException
import br.com.ntconsult.challenge.model.Session
import br.com.ntconsult.challenge.repository.SessionRepository
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

@Service
class SessionService(private val sessionRepository: SessionRepository) {

    fun create(sessionDefinition: SessionDefinition) : Session {
        if(sessionRepository.existSessionForTopic(sessionDefinition.topic.id!!)) {
            throw SessionAlreadyExistsException("There is already a session for topic {ID} = ${sessionDefinition.topic.id}")
        }
        return with(sessionDefinition) {
            val now = LocalDateTime.now()
            val session = Session(
                    topic = topic,
                    startedAt = now,
                    finishAt = now.plusMinutes(timeToVote)
            )
            save(session)
        }
    }

    fun save(session: Session): Session {
        return sessionRepository.save(session)
    }

    @Cacheable("sessions")
    fun findByTopicId(id: String) : Session? {
        return sessionRepository.findByTopicId(UUID.fromString(id))
    }

    fun findAllSessionsClosedWithTopicResultEmpty(date: LocalDateTime): List<Session> {
        return sessionRepository.findAllSessionsClosedWithResultDetailEmpty(date)
    }

}